spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessor.java

```diff
@@ -574,14 +575,13 @@ else if (exchangeType.equals(ExchangeTypes.HEADERS)) {
 					}
 				}
 				else {
-					Assert.isTrue(value instanceof String, "value must resolve to a String for type conversion");
-					String valueString = (String) value;
-					if (StringUtils.hasText(valueString)) {
-						if (CONVERSION_SERVICE.canConvert(String.class, typeClass)) {
-							map.put(key, CONVERSION_SERVICE.convert(valueString, typeClass));
+					if (value instanceof String && StringUtils.hasText((String) value)) {
+						if (CONVERSION_SERVICE.canConvert(value.getClass(), typeClass)) {
+							map.put(key, CONVERSION_SERVICE.convert(value, typeClass));
 						}
 						else {
-							throw new IllegalStateException("Cannot convert from String to " + typeClass);
+							throw new IllegalStateException("Cannot convert from " + value.getClass()
+								+ " to " + typeClass);
```

Comment:
Yeah... Looks like we have to use `Class.getName()` here as well.
Although I can live with this extra `class` literal in the exception message.